_posts/2024-03-30-nsb-asb-retries.md

```diff
@@ -0,0 +1,63 @@
+
+However, if the NServiceBus endpoint process crashes completely (e.g. due to an OutOfMemory exception), the recoverability pipeline can't be invoked anymore. Instead, the message will become available for processing again (the exact behavior depends on the specific transport, e.g. with Azure Service Bus the lease lock will expire and the message will be available for consumers again). If the endpoint process always crashes on a specific message content, the recoverability features will never have the chance to interrupt this behavior, potentially leading to infinite retries.
+
+Note: The NServiceBus recoverability pipeline can't be invoked preventively as the pipeline requires the exception details to determine the proper recoverability action. Some transports might store the exception information in memory to deal with transaction limitations (e.g. [MSMQ transport](https://github.com/Particular/NServiceBus.Transport.Msmq/blob/master/src/NServiceBus.Transport.Msmq/SendsAtomicWithReceiveNativeTransactionStrategy.cs#L84)) but such approaches also won't work in the described scenario as a process crash also wipes any temporary storage.
+
+
+## Solutions
+
+To prevent infinite retries of messages, the message processing must be interrupted before executing the handler which crashes the endpoint process. There is no built-in solution in NServiceBus but depending on the transport used we can build this ourselves. When using Azure Service Bus transport, this is not too difficult to achieve, as Azure Service Bus already tracks the delivery count of a message internally. We can make use of this in 2 ways:
+
+### ASB native DLQ
+
+Azure Service Bus already comes with a built-in dead-letter mechanism which can be configured using the queue's *max delivery count* setting. When the message has been delivered (but not consumed) to a client more than the configured amount, the message is moved to the dead-letter queue.
+
+When you let NServiceBus create your ASB queues, this value is set to `int.MaxValue` to prevent it from interfering with the NServiceBus recoverability configuration. You can take control of the queue creation yourself and set a more restrictive message processing limit on the queues. This is a simple solution but has some disadvantages:
+
+* NServiceBus doesn't allow the configuration of this value, not via [code-first configuration](https://docs.particular.net/transports/azure-service-bus/configuration) nor the [CLI tool](https://docs.particular.net/transports/azure-service-bus/operational-scripting). Therefore you need to create the queues with all the other required settings manually which can be error-prone.
+* The message is moved to the Azure Service Bus dead-letter queue instead of the NServiceBus configured error queue which makes it more difficult to integrate with ServiceControl.
```

Comment:
```suggestion
* NServiceBus doesn't allow the configuration of this value, not via [code-first configuration](https://docs.particular.net/transports/azure-service-bus/configuration) or the [CLI tool](https://docs.particular.net/transports/azure-service-bus/operational-scripting). Therefore, you need to create the queues with all the other required settings manually, which can be error-prone.
* The message is moved to the Azure Service Bus dead-letter queue instead of the NServiceBus configured error queue, making it more challenging to integrate with ServiceControl.
```